core/src/services/zookeeper/backend.rs

```diff
@@ -0,0 +1,246 @@
+        self
+    }
+
+    /// Specify the auth digest path of zookeeper service
+    pub fn digest_path(&mut self, digest_path: &str) -> &mut Self {
+        if !digest_path.is_empty() {
+            self.digest_path = Some(digest_path.to_string());
+        }
+        self
+    }
+}
+
+impl Debug for ZookeeperBuilder {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Builder");
+        if let Some(endpoint) = self.endpoint.clone() {
+            ds.field("endpoint", &endpoint);
+        }
+        if let Some(user) = self.user.clone() {
+            ds.field("user", &user);
+        }
+        if let Some(digest_path) = self.digest_path.clone() {
+            ds.field("digest_path", &digest_path);
+        }
+        ds.finish()
+    }
+}
+
+impl Builder for ZookeeperBuilder {
+    const SCHEME: Scheme = Scheme::Zookeeper;
+    type Accessor = ZookeeperBackend;
+
+    fn from_map(map: HashMap<String, String>) -> Self {
+        let mut builder = ZookeeperBuilder::default();
+
+        map.get("endpoint").map(|v| builder.endpoint(v));
+        map.get("user").map(|v| builder.user(v));
+        map.get("digest_path").map(|v| builder.digest_path(v));
+
+        builder
+    }
+
+    fn build(&mut self) -> Result<Self::Accessor> {
+        let endpoint = match self.endpoint.clone() {
+            None => DEFAULT_ZOOKEEPER_ENDPOINT.to_string(),
+            Some(endpoint) => endpoint,
+        };
+        let (auth, acl) = match (self.user.clone(), self.digest_path.clone()) {
+            (Some(user), Some(digest_path)) => {
+                let password = std::fs::read(digest_path).map_err(parse_file_error)?;
+                let auth = [format!("{user}:").as_bytes().to_vec(), password].concat();
+                (auth, zk::Acl::creator_all())
+            }
+            _ => {
+                // TODO: change to warn
+                warn!("username and password isn't set, default use `anyone` acl");
+                (Vec::<u8>::new(), zk::Acl::anyone_all())
+            }
+        };
+
+        Ok(ZookeeperBackend::new(ZkAdapter {
+            endpoint,
+            auth,
+            acl,
+            client: OnceCell::new(),
+        }))
+    }
+}
+
+/// Backend for Zookeeper service
+pub type ZookeeperBackend = kv::Backend<ZkAdapter>;
+
+#[derive(Clone)]
+pub struct ZkAdapter {
+    endpoint: String,
+    auth: Vec<u8>,
+    client: OnceCell<zk::Client>,
+    acl: &'static [zk::Acl],
+}
+
+impl Debug for ZkAdapter {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        let mut ds = f.debug_struct("Adapter");
+        ds.field("endpoint", &self.endpoint);
+        ds.field("acl", &self.acl);
+        ds.finish()
+    }
+}
+
+impl ZkAdapter {
+    async fn get_connection(&self) -> Result<zk::Client> {
+        if let Some(client) = self.client.get() {
+            return Ok(client.clone());
+        }
+        match zk::Client::connect(&self.endpoint.clone()).await {
```

Comment:
Seems we can't recover from broken connection. How about using bb8 to do this? We can take `memecache` service as a reference.